File: src/ips.rs
use crate::utils::*;

use std::fs::write;
use std::path::Path;

const OFFSET_SIZE: usize = 3;
const SIZE_SIZE: usize = 2;

const IPS_HEADER: &[u8] = b"PATCH";
const IPS_FOOTER: &[u8] = b"EOF";

pub fn patch_ips<P: AsRef<Path>>(src_file: &P, patch_file: &P, out_file: &P) -> Result<(), &'static str> {
    let src = read_file(src_file);
    let patch = read_file(patch_file);

    let out = apply_ips(&src, &patch);
    match out {
        Ok(dst) => {
            write(out_file, dst).expect("Unable to open output file");
            Ok(())
        },
        Err(e) => {
            Err(e)
        }
    }
}

fn apply_ips(src: &Vec<u8>, patch: &[u8]) -> Result<Vec<u8>, &'static str> {
    let mut target = src.clone();

    // Verify that this patch file is indeed valid IPS
    for (i, &byte) in IPS_HEADER.iter().enumerate() {
        if patch[i] != byte {
            return Err("This is not a valid IPS patch file");
        }
    }

    let footer_offset = patch.len() - IPS_FOOTER.len();
    for (i, &byte) in IPS_FOOTER.iter().enumerate() {
        if patch[i + footer_offset] != byte {
            return Err("This is not a valid IPS patch file");
        }
    }

    let mut idx = IPS_FOOTER.len();
    loop {
        if idx == footer_offset {
            break;
        } else if idx > footer_offset {
            return Err("End of file was overrun");
        }

        let mut rle = false;
        let offset = pack_u24(&patch[idx..(idx + OFFSET_SIZE)]);
        idx += OFFSET_SIZE;
        let mut size = pack_u16(&patch[idx..(idx + SIZE_SIZE)]);
        idx += SIZE_SIZE;

        if size == 0 {
            rle = true;
            size = pack_u16(&patch[idx..(idx + SIZE_SIZE)]);
            idx += SIZE_SIZE;
        }

        if offset + size > target.len() {
            return Err("This patch file is corrupted");
        }

        if rle {
            let data = patch[idx];
            for i in 0..size {
                target[offset + i] = data;
            }
            idx += 1;
        } else {
            for i in 0..size {
                target[offset + i] = patch[idx + i];
            }
            idx += size;
        }
    }

    Ok(target)
}

File: src/ips.nim
import streams
import os

import err
import utils

const OFFSET_LEN = 3
const SIZE_LEN = 2
const IPS_HEADER = "PATCH"
const IPS_FOOTER = "EOF"
const BUFFER_SIZE = 0xFFFF

proc is_ips*(patch: FileStream, length: int): bool =
    let header = patch.readStr(IPS_HEADER.len())
    patch.setPosition(length - IPS_FOOTER.len())
    let footer = patch.readStr(IPS_FOOTER.len())
    patch.setPosition(0)
    return header == IPS_HEADER and footer == IPS_FOOTER

proc patch_ips*(patch: FileStream, patch_len: int, source_name, output_name: string) =
    if not source_name.fileExists():
        raise newException(IOError, "Source file does not exist")

    # TODO: May need to check for directory permissions?
    copyFile(source_name, output_name)
    var output = open(output_name, FileMode.fmReadWriteExisting)
    let source_len = int(source_name.getFileSize())
    var buffer: array[BUFFER_SIZE, uint8]

    patch.setPosition(IPS_HEADER.len())
    let footer_offset = patch_len - IPS_FOOTER.len()
    while true:
        let idx = patch.getPosition()
        if idx == footer_offset:
            break
        elif idx > footer_offset:
            raise newException(CorruptionError, "End of file was overrun")

        let offset = patch.readUintX(OFFSET_LEN)
        var size = patch.readUintX(SIZE_LEN)

        var rle = false
        if size == 0:
            rle = true
            size = patch.readUintX(SIZE_LEN)

        if offset + size > source_len:
            raise newException(CorruptionError, "This patch file is corrupted")

        output.setFilePos(offset)
        if rle:
            let data = patch.readUint8()
            for i in 0..<size:
                buffer[i] = data
            discard output.writeBytes(buffer, 0, size)
        else:
            discard patch.readData(addr(buffer), size)
            discard output.writeBytes(buffer, 0, size)

    output.close()

